### ISSUE TITLE
SSH/OpenVPN port forwarding behind protected network for incoming connection

### ISSUE THREAD

**Issue by Hideman85** [BiglySoftware/BiglyBT#2006]
Hi community,

I have an issue with my network NAT is not working/blocked and so I can't open port for incoming connection. I'm still able to open connection to peers but I'm limited and so I have only access to 10% of the available peers.

I try to find out the config to for incoming connection (interface/IP & port to bind) but I'm not able to do it.

Basically I try to use [this free service](portmap.io) I just found to open a forwarded port on my machine
```
sudo openvpn --config config.ovpn
2021-02-19 20:10:50 [193.161.193.99] Peer Connection Initiated with [AF_INET]193.161.193.99:****
2021-02-19 20:10:51 TUN/TAP device tun0 opened
2021-02-19 20:10:51 net_iface_mtu_set: mtu 1500 for tun0
2021-02-19 20:10:51 net_iface_up: set tun0 up
2021-02-19 20:10:51 net_addr_ptp_v4_add: 10.8.35.50 peer 10.8.35.49 dev tun0

ifconfig
tun0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
        inet 10.8.35.50  netmask 255.255.255.255  destination 10.8.35.49
        unspec 00-00-00-00-00-00-00-00-00-00-00-00-00-00-00-00  txqueuelen 500

nc -vz 193.161.193.99 *****
193.161.193.99: inverse host lookup failed: Unknown host
(UNKNOWN) [193.161.193.99] ***** (openvpn) open
```
Now if I understand well I have a mapping port from the remote server and I should listen to the local port over this tunnel (port given by the site)

I just don't find how to change this config in BiglyBT

Basically this solution would unlock people in firewall protected network or like me where my provider use Carrier Grade NAT...

**Comment by magicmana** [comment]
hello....
On firewall protected network; the  best solution is  : to not use a  bittorrent client....      

If you have No access on you  home router  like with an  web interface for open port .. had you try to enable UpNP , 
most routers provided by internet service providers for home; support it.

on the webpage   on https://portmap.io/  they  tell "Random external port"  for  free access.  that want  tell    each time  you start   your vpn    , you need to  grep the  port and  past it  in biglybt setting , and  restart  the client.

an   finally   if your home  router is  not a "basic"  router    have chance  he have a   anti ip-spoofing   rules  too ....
and   drop all  your packet...

 https://en.wikipedia.org/wiki/IP_address_spoofing

**Comment by Hideman85** [comment]
> On firewall protected network; the best solution is : to not use a bittorrent client....

- What a useful comment...


> If you have No access on you home router like with an web interface for open port .. had you try to enable UpNP ,

- The reply is in my previous post:
  > Basically this solution would unlock people in firewall protected network or like me where my provider use **Carrier Grade NAT...**


> On the webpage on https://portmap.io/ they tell "Random external port" for free access

- If you have a deeper look you will find out that they give you a random port **Once** and it does not change further more...


> an finally if your router is not a home "basic" router ; [...]  have a anti ip-spoofing rules [...] packet will be dropped

- At least do you know about OpenVPN / SSH tunneling?

**Comment by magicmana** [comment]
personally   my  job should   be so  much  easy ,  if  the user  of's  my's  firewall's  if  employee  not   installing   inapropritate software on  their  professional computer .... 

and i know   well  vpn SSL and  IPsec... i'm  certified on all Fortinet  Firewall product...


for  your  information  a VPN  (ipsec or ssl)  not a  "magical"  2 way protocole....
the port  are  used  by ip address who  use route table to communicate...



i will  take a small  sample..
1) a public IP address  (100.100.100.100 for the sample) sent a TCP  packet on the   public  ip  public ip of  your VPN (200.200.200.200) on port 5000.
2)  the public  VPN  send  the packet   to  your   private  IP   "10.8.35.50"  (the ip in   your  log)  from the address "10.8.35.49"

3)  your packet arrive   in  biglybt  from  the  public address  100.100.100.100.



4) now  if  biglybt  want  to  respond    to "100.100.100.100"  il will  interrogate  your computer  route  table.
your  computer  respond "100.100.100.100 ?   not  listed on my route table  >> use  the  default  gataway route  192.168.0.254..."  
 
5) your local (192.168.0.254)  router  receive  the packet   from  biglybt .  he  interrogate  his  routing  table who  tell  : 
send all packet  to 10.99.0.254 (your private network  from   your  routeur to your  ISP  gateway...

6) the packet arrive  on your ISP  NAT  server on address 10.99.0.254 ...  and    if is  a decent  router  like a cisco router well configured ;   the  NAT router  tell :  
"stange...  i  receive a TCP session  packet  for the address "100.100.100.100"  but   haven't  see   TCP session   from "100.100.100.100"  and i haven't  "100.100.100.100"  in my NAT  table    Should be a good idea to drop the packet".


the  only  way for  bypass  that is to use  VPN      ,  and  who  route all  your   internet  traffic  on  the  VPN  and who create a port  forwarding  at  the same time...

   

 

and last , 
for  maximise   the  discovery on  other pair  in a  bittorrent  client    ,   you  need   use 2 or 3 ports ....  1 port  TCP  one port  UDP  for  your  client.   and  if  you want  have the possibility to  receive the peer list  directy from other client , you  need  enable DHT and use one more UDP port.

**Comment by Hideman85** [comment]
Hi @magicmana,

I understand better what you wanted to mean with the IP spoofing and appreciate your detailed reply :)

I know the easiest solution is to VPN all your traffic but either you have to pay for having a high speed VPN or either you suffer from a lower speed.

What I want is to not reduce my outgoing (since nothing is block) but find a way to enable incoming connection without having the possibility to obtain a public IP (yes that's a shame but this is how it is...) that's why I'm trying to tunnel only inbound connection (and there reply of course).

The issue is some of my download are stalled because I can't connect peers (that are probably in the same situation as me (blocked incoming) since I checked with nmap and port is seen as closed) and they can't connect to me since my incoming is also blocked... So at the end we only know that we exists :D

(Note: It does not mean that if I enable incoming, my torrent would be able to continue (but at least I would know the state of these peers))

If you know a nice way to solve that challenge I would be happy to learn it ;)

**Comment by magicmana** [comment]
have more harder challenge.... ;-)
if i have this  problem   in the  futur  a  solution i could  be deploy  sould be :
-install  a   virtual  machine (or use another computer )
-on the  vitual machine  install the  the cheapest vpn i  can found  with one or more port  forwarding
- install a  client bittorrent  who  support   the sharing of  peer on local network; and who can be managed by command line/script 
-set the  traffic shipping on the client   bittorrent installed on the virtual machine at 1ko/s for upload/download. for not eating the bandwidth of VPN
-enable  the  "peer sharing  on  local  network"  on the both  client.
- configure a directory for auto add torrent on  the virtual machine  bittorent client
- make a script like a rsync with a cron every min  to ,  duplicate  all   torrent from  the your local computer to the virtual machine
- and finally  make a script  for delete the torrent on  the virtual  machine client who have more than  xxx hour/day...
